drivers: video: formats: add most of the raw formats

josuah opened the pull request:
Downstream:

- https://github.com/zephyrproject-rtos/zephyr/pull/88011

This adds the raw formats in the same packing as defined by MIPI, with the naming that follows the Linux Kernel:

- Grayscale: https://docs.kernel.org/userspace-api/media/v4l/pixfmt-yuv-luma.html
- Bayer: https://docs.kernel.org/userspace-api/media/v4l/pixfmt-bayer.html

josuah committed:
drivers: video: formats: include bayer 10/12/14-bits formats

In addition to the 8-bit, introduce all the other bayer formats
described by MIPI-CSI2 specification. The 8-bit bayer formats
description is shortened to just 4 bytes like the other formats,
to help intuition while comparing the different formats.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: formats: add grayscale formats

Add the grayscale formats in the same packing as defined by MIPI.
The Linux V4L2 naming is preserved.

Signed-off-by: Josuah Demangeon <me@josuah.net>

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -1005,6 +1005,58 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Yyyyyyyy | Yyyyyyyy | Yyyyyyyy | Yyyyyyyy | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_Y8 VIDEO_FOURCC('G', 'R', 'E', 'Y')
```
It seems confused between : [_PIX_FMT_Y8I](https://elixir.bootlin.com/linux/v6.11.1/source/include/uapi/linux/videodev2.h#L798) and [_PIX_FMT_GREY](https://elixir.bootlin.com/linux/v6.11.1/source/include/uapi/linux/videodev2.h#L590).

I think it should be:

`#define VIDEO_PIX_FMT_GREY VIDEO_FOURCC('G', 'R', 'E', 'Y')`

josuah replied:
I sometimes have difficulties making a difference between what Linux status-quo and what Linux tries to move towards. I was borrowing Y8 from [media bus](https://docs.kernel.org/userspace-api/media/v4l/subdev-formats.html#id9).

I added a hint in the documentation that this is nothing special, just following the FOURCC naming.

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | rrggrrgg | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SBGGR10P VIDEO_FOURCC('p', 'R', 'A', 'A')
```
FOURCC seems wrong, should be:

`#define VIDEO_PIX_FMT_SBGGR10P VIDEO_FOURCC('p', 'B', 'A', 'A')`

ref : https://elixir.bootlin.com/linux/v6.11.1/source/include/uapi/linux/videodev2

josuah replied:
```diff
-#define VIDEO_PIX_FMT_SBGGR10P VIDEO_FOURCC('p', 'R', 'A', 'A')
+#define VIDEO_PIX_FMT_SBGGR10P VIDEO_FOURCC('p', 'B', 'A', 'A')
```
ok

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | ggrrggrr | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SGBRG10P VIDEO_FOURCC('p', 'g', 'A', 'A')
```
FOURCC seems wrong too, should be:

`#define VIDEO_PIX_FMT_SGBRG10P VIDEO_FOURCC('p', 'G', 'A', 'A')`

josuah replied:
```diff
-#define VIDEO_PIX_FMT_SGBRG10P VIDEO_FOURCC('p', 'g', 'A', 'A')
+#define VIDEO_PIX_FMT_SGBRG10P VIDEO_FOURCC('p', 'G', 'A', 'A')
```
ok

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | ggbbggbb | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SGRBG10P VIDEO_FOURCC('p', 'G', 'A', 'A')
```
FOURCC seems wrong too (swapped with the above), should be:

`#define VIDEO_PIX_FMT_SGRBG10P VIDEO_FOURCC('p', 'g', 'A', 'A')`

josuah replied:
```diff
-#define VIDEO_PIX_FMT_SGRBG10P VIDEO_FOURCC('p', 'G', 'A', 'A')
+#define VIDEO_PIX_FMT_SGRBG10P VIDEO_FOURCC('p', 'g', 'A', 'A')
```
ok

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | bbggbbgg | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SRGGB10P VIDEO_FOURCC('p', 'B', 'A', 'A')
```
same, should be:

#define VIDEO_PIX_FMT_SRGGB10P VIDEO_FOURCC('p', 'R', 'A', 'A')

josuah replied:
```diff
-#define VIDEO_PIX_FMT_SRGGB10P VIDEO_FOURCC('p', 'B', 'A', 'A')
+#define VIDEO_PIX_FMT_SRGGB10P VIDEO_FOURCC('p', 'R', 'A', 'A')
```
ok

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Gggggggg | Rrrrrrrr | rrrrgggg | Gggggggg | Rrrrrrrr | rrrrgggg | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SBGGR12P VIDEO_FOURCC('p', 'R', 'C', 'C')
```
FOURCC : pBCC

josuah replied:
```diff
-#define VIDEO_PIX_FMT_SBGGR12P VIDEO_FOURCC('p', 'R', 'C', 'C')
+#define VIDEO_PIX_FMT_SBGGR12P VIDEO_FOURCC('p', 'B', 'C', 'C')
```
ok

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Rrrrrrrr | Gggggggg | ggggrrrr | Rrrrrrrr | Gggggggg | ggggrrrr | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SGBRG12P VIDEO_FOURCC('p', 'g', 'C', 'C')
```
pGCC

josuah replied:
```diff
-#define VIDEO_PIX_FMT_SGBRG12P VIDEO_FOURCC('p', 'g', 'C', 'C')
+#define VIDEO_PIX_FMT_SGBRG12P VIDEO_FOURCC('p', 'G', 'C', 'C')
```
ok

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Bbbbbbbb | Gggggggg | ggggbbbb | Bbbbbbbb | Gggggggg | ggggbbbb | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SGRBG12P VIDEO_FOURCC('p', 'G', 'C', 'C')
```
pgCC

josuah replied:
```diff
-#define VIDEO_PIX_FMT_SGRBG12P VIDEO_FOURCC('p', 'G', 'C', 'C')
+#define VIDEO_PIX_FMT_SGRBG12P VIDEO_FOURCC('p', 'g', 'C', 'C')
```
ok

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Gggggggg | Bbbbbbbb | bbbbgggg | Gggggggg | Bbbbbbbb | bbbbgggg | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SRGGB12P VIDEO_FOURCC('p', 'B', 'C', 'C')
```
pRCC

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | rrgggggg | ggggrrrr | rrrrrrgg | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SBGGR14P VIDEO_FOURCC('p', 'R', 'E', 'E')
```
pBEE

josuah replied:
```diff
-#define VIDEO_PIX_FMT_SBGGR14P VIDEO_FOURCC('p', 'R', 'E', 'E')
+#define VIDEO_PIX_FMT_SBGGR14P VIDEO_FOURCC('p', 'B', 'E', 'E')
```
ok

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | ggrrrrrr | rrrrgggg | ggggggrr | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SGBRG14P VIDEO_FOURCC('p', 'g', 'E', 'E')
```
pGEE

josuah replied:
```diff
-#define VIDEO_PIX_FMT_SGBRG14P VIDEO_FOURCC('p', 'g', 'E', 'E')
+#define VIDEO_PIX_FMT_SGBRG14P VIDEO_FOURCC('p', 'G', 'E', 'E')
```
ok

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | ggbbbbbb | bbbbgggg | ggggggbb | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SGRBG14P VIDEO_FOURCC('p', 'G', 'E', 'E')
```
pgEE

josuah replied:
```diff
-#define VIDEO_PIX_FMT_SGRBG14P VIDEO_FOURCC('p', 'G', 'E', 'E')
+#define VIDEO_PIX_FMT_SGRBG14P VIDEO_FOURCC('p', 'g', 'E', 'E')
```
ok

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -856,41 +856,155 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | bbgggggg | ggggbbbb | bbbbbbgg | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_SRGGB14P VIDEO_FOURCC('p', 'B', 'E', 'E')
```
pREE

josuah replied:
```diff
-#define VIDEO_PIX_FMT_SRGGB14P VIDEO_FOURCC('p', 'B', 'E', 'E')
+#define VIDEO_PIX_FMT_SRGGB14P VIDEO_FOURCC('p', 'R', 'E', 'E')
```
ok

josuah commented:
Sorry about that, I think I copy-pasted the RGGB8 etc formats from Zephyr and looked at the [documentation](https://docs.kernel.org/userspace-api/media/v4l/pixfmt-srggb10p.html) (in opposite order) for adding the FOURCC... I could have checked better... Thank you for the extra pair of vigilant eyes.

josuah commented:
Force-push:

- Added a hint that GREY was the same as "Y8" to help users coming from other libraries (i.e. LVGL calls it L8)

josuah committed:
drivers: video: improve the format description docs

@verbatim shows the leading comment '*' fence in the output.
@code{.unparsed} allows extracting the text only.
In some formats, the pixels are also not immediately packed into fixed
number of bytes, like YUYV due to chroma subsampling. Disambiguate.
Add numbers marks on top to help identify the individual pixels.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Sorry one more doc typo! I had the wrong bit endianness for RAW10's continuation byte: '0  1  2  3' while it should have been `3  2  1  0`.

I took this opportunity to dust it a bit too. `@verbatim` -> `@code` to avoid a glitch, etc.

Before: https://docs.zephyrproject.org/latest/doxygen/html/group__video__pixel__formats.html#ga64ad74bb6c92041a4190614b684ae024
after: https://builds.zephyrproject.io/zephyr/pr/88138/docs/doxygen/html/group__video__pixel__formats.html#ga64ad74bb6c92041a4190614b684ae024
